### PR TITLE
Faces

### PR DESCRIPTION
**Pull request by ricardofantin**
Branch faces update with code from master.

Only the meson build works. The autotools checks for opencv, but doesn't install yet.

There are others bugs I want to work in bountysource, I will need everything finished until 2018/02/20 to get the the bounty without paying fees. Sorry to ask, but close to that date can we make a haste to merge things?

The warning that existed in the first pull-request some time ago have been solved.

To compile use "meson builddir/ -Dface=true"

Thanks.

### CODE REVIEW

**Review by phako — COMMENTED** [review]
Meh. That branch is an awfull mess. Its age doesn't really help it. I would much prefer to have it rebased on the current master, tough I don't know how to keep the proper attribution

**Comment by phako** [comment]
There is an issue with src/Dialogs.vala:

```
r.vapi src/plugins/shotwell-plugin-dev-1.0.vapi
../src/Dialogs.vala:661.1-661.2: error: syntax error, expected identifier
<<<<<<< HEAD
^^
../src/Dialogs.vala:984.1-984.2: error: syntax error, expected identifier
=======
^^
../src/dialogs/EntryMultiCompletion.vala:10.1-10.55: error: The root namespace already contains a definition for `EntryMultiCompletion'
public class EntryMultiCompletion : Gtk.EntryCompletion {
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../src/Dialogs.vala:664.1-664.55: note: previous definition of `EntryMultiCompletion' was here
public class EntryMultiCompletion : Gtk.EntryCompletion {
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../src/dialogs/MultiTextEntryDialog.vala:9.1-9.46: error: The root namespace already contains a definition for `MultiTextEntryDialog'
public class MultiTextEntryDialog : Gtk.Dialog {
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../src/Dialogs.vala:932.1-932.46: note: previous definition of `MultiTextEntryDialog' was here
public class MultiTextEntryDialog : Gtk.Dialog {
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../src/dialogs/SetBackgroundSlideshow.vala:9.1-9.54: error: The root namespace already contains a definition for `SetBackgroundSlideshowDialog'
public class SetBackgroundSlideshowDialog : Gtk.Dialog {
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../src/Dialogs.vala:788.1-788.54: note: previous definition of `SetBackgroundSlideshowDialog' was here
public class SetBackgroundSlideshowDialog : Gtk.Dialog {
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../src/dialogs/SetBackground.vala:9.1-9.50: error: The root namespace already contains a definition for `SetBackgroundPhotoDialog'
public class SetBackgroundPhotoDialog : Gtk.Dialog {
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../src/Dialogs.vala:754.1-754.50: note: previous definition of `SetBackgroundPhotoDialog' was here
public class SetBackgroundPhotoDialog : Gtk.Dialog {
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../src/dialogs/TextEntry.vala:9.1-9.41: error: The root namespace already contains a definition for `TextEntryDialog'
public class TextEntryDialog : Gtk.Dialog {
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../src/Dialogs.vala:862.1-862.41: note: previous definition of `TextEntryDialog' was here
public class TextEntryDialog : Gtk.Dialog {
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Compilation failed: 7 error(s), 0 warning(s)

```

File: src/AppDirs.vala [hunk 1]
```diff
@@ -326,6 +326,9 @@ class AppDirs {
     public static File get_facedetect_bin() {
         const string filename = "shotwell-facedetect";
         File f = File.new_for_path(AppDirs.get_exec_dir().get_path() + "/facedetect/" + filename);
+        if (!f.query_exists()) {//testing meson builddir
+            f = File.new_for_path(AppDirs.get_exec_dir().get_path() + "/../facedetect/" + filename);
+        }
```
Comment (phako) [inline src/AppDirs.vala]: Please use the File api to navigate the tree, eg.g. get_exec_dir().get_parent().get_child("facedetect") etc. or resolve_relative_path(). I know the rest of the code doesn't do it, but we should start to get a bit nicer here

File: meson.build [hunk 2]
```diff
@@ -81,6 +81,7 @@ if get_option('disable-dupe-detection')
   add_global_arguments(['--define=NO_DUPE_DETECTION'], language : vala)
 endif
 
+face_resources = (['main.vala']) #hack to avoid error when no face option used
```
Comment (phako) [inline meson.build]: Should be possible to use just []. But please move the definition of the faces things into src/meson.build

**Comment by ricardofantin** [comment]
Hello,
the code is divided in two two folders facedetec/ and src/faces . The facedetec is written in C++, if I remove this needed and use only vala the code will be only in src/faces and the code will be less messy. I can move all definition to src/meson.build easily if I use only vala.
The compilation error was my error in commit, I will correct it.
In
 +face_resources = (['main.vala']) #hack to avoid error when no face option used 
I can't put only [] because a error appear about a file with no name not found.

I will change the needed of C++ and others things, but it will take a while.

**Comment by phako** [comment]
What I mean by mess is not the code but the history. I have serious issues to merge all that properly. I would much prefer a more linear history of things.

Please do not worry about C++. It is fine. Really.

using [] works fine here when doing this:
iff --git a/meson.build b/meson.build
index 0cb8c05e..122a94cb 100644
--- a/meson.build
+++ b/meson.build
@@ -81,11 +81,11 @@ if get_option('disable-dupe-detection')
   add_global_arguments(['--define=NO_DUPE_DETECTION'], language : vala)
 endif
 
-face_resources = (['main.vala']) #hack to avoid error when no face option used
+face_resources = []
 if get_option('face')
   add_global_arguments(['--define=ENABLE_FACES'], language : 'vala')
   subdir('facedetect')
-  face_resources = (['faces/FacesBranch.vala', 'faces/FaceLocation.vala', 'faces/FacePage.vala', 'faces/FaceShape.vala', 'faces/Faces.vala', 'faces/Face.vala', 'db/FaceLocationTable.vala', 'db/FaceTable.vala'])
+  face_resources += ['faces/FacesBranch.vala', 'faces/FaceLocation.vala', 'faces/FacePage.vala', 'faces/FaceShape.vala', 'faces/Faces.vala', 'faces/Face.vala', 'db/FaceLocationTable.vala', 'db/FaceTable.vala']
 endif
 
 json_glib = dependency('json-glib-1.0')

Have you tried compilin the branch with -Dface=false? that doesn't seem to work.

**Comment by phako** [comment]
Also, please install the facedetect binary to libexec path, similar to the thumbnailer. In parallel I am trying to find out why the face detection is done in an external helper binary.

**Comment by phako** [comment]
haarcascade_frontalface_alt2.xml and haarcascade_frontalface_default.xml are HTML files from github, not the actual XML files.

**Comment by phako** [comment]
Can you please continue working on the wip/faces-rebased branch? Also, please do not merge master into the branch, but rebase the branch on top of master, please?

**Comment by ricardofantin** [comment]
I will continue working on this branch. (I actually want to implement face recognition, some day).
I never used the rebase option and didn't know what is it. I will take a look in it.

I'm on work now, tonight I will take a look and make the correct commit to everything compile.

**Comment by phako** [comment]
Technically it's not a rebase, more applying a cleaned diff of your branch against master on top of master. I think I fixed the compile issues while doing that.

**Comment by ricardofantin** [comment]
Every modification suggestion were made.

Some problems are:
If use
DESTDIR=$HOME ninja install 
The facedetec executable and the facedetect-haarcascade.xml will not be found. The code only looks form meson build dir and installation build dir.
I don't think it is a major issue.

The Face menu population use string's from src/Resources.vala . I don't know how the translation works. May be some modification will be needed.

I would like to exclude the 9d29207a09475032f6d7523929f27dec2bddfed5 commit. It included tow opencv face description to the code. Is there a way to remove old git commits?

**Comment by phako** [comment]
> I would like to exclude the 9d29207 commit. It included tow opencv face description to the code. Is there a way to remove old git commits?

Usually you just skip it during an interactive git rebase

**Comment by phako** [comment]
Superseeded by PR #3

**Pull request closed without merging**